### PR TITLE
Linked wallet addresses

### PR DESCRIPTION
**Pull request by soloveyhappy**
Adding functionality of linking additional user wallet addresses with the master wallet. No matter from what wallet user make operations from master or from linked.

### CODE REVIEW

File: smartcontracts/CryptaurDepository.sol [hunk 1]
```diff
@@ -61,7 +60,124 @@ contract CryputarReserveFund {
...
+
+    function isMasterWallet(address addr) internal constant returns (bool) {
+        return masterToSlaves[addr].hasValue;
+    }
+
+    function isLinkedWallet(address addr) internal constant returns (bool) {
+        return slaveToMasterAddress[addr] != address(0);
+    }
+
+    /// Guess that address book already had chnging address
```
Comment (isvirin) [inline smartcontracts/CryptaurDepository.sol]: * we don't use "///". all comments must be in doxygen style
* typo "chnging"

File: smartcontracts/CryptaurDepository.sol [hunk 2]
```diff
@@ -61,7 +60,124 @@ contract CryputarReserveFund {
...
+        address[] values;
+        mapping(address => uint) keys;
+    }
+
+    event WalletLinked(address indexed master, address indexed slave);
+    event WalletUnlinked(address indexed master, address indexed slave);
+    event AddressChanged(address indexed _old, address indexed _new);
+
+    mapping (address => AddressRelations) private masterToSlaves;
+    mapping(address => address) private slaveToMasterAddress;
```
Comment (isvirin) [inline smartcontracts/CryptaurDepository.sol]: coding style "mapping (" or "mapping(" - space?

File: smartcontracts/CryptaurDepository.sol [hunk 3]
```diff
@@ -61,7 +60,124 @@ contract CryputarReserveFund {
...
+    }
+
+    struct SlaveDictionary
+    {
+        address[] values;
+        mapping(address => uint) keys;
+    }
+
+    event WalletLinked(address indexed master, address indexed slave);
+    event WalletUnlinked(address indexed master, address indexed slave);
```
Comment (isvirin) [inline smartcontracts/CryptaurDepository.sol]: coding-style! we use "_" prefix in functions and events parameters

File: smartcontracts/CryptaurDepository.sol [hunk 4]
```diff
@@ -107,96 +227,130 @@ contract CryptaurDepository is owned {
...
+    function setMaxLinkedWalletCount(uint8 newMaxCount) public {
+        require(msg.sender == owner || msg.sender == backend);
+        maxLinkedWalletCount = newMaxCount;
     }
     
     function freeze(address _who, bool _freeze) onlyOwner public {
-        freezed[_who] = _freeze;
-        Freeze(_who, _freeze);
+        address masterWallet = getMasterWallet(_who);
+        if (masterWallet != address(0))
```
Comment (isvirin) [inline smartcontracts/CryptaurDepository.sol]: Are you sure in this condition??? I think we need to change "!=" to "=="

File: smartcontracts/CryptaurDepository.sol [hunk 5]
```diff
@@ -107,96 +227,130 @@ contract CryptaurDepository is owned {
...
+        Deposit(masterWallet, _amount, _txHash);     
     }
     
     function withdraw(uint _amount) public {
-        require(balanceOf[msg.sender] >= _amount);
-        require(!freezed[msg.sender]);
-        balanceOf[msg.sender] -= _amount;
-        ERC20Base(cryptaurToken).transfer(msg.sender, _amount);
-        Withdraw(msg.sender, _amount);
+        address masterWallet = getMasterWallet(msg.sender);
```
Comment (isvirin) [inline smartcontracts/CryptaurDepository.sol]: what will happens if masterWallet will be zero?

Comment (soloveyhappy) [reply smartcontracts/CryptaurDepository.sol]: in that case next require will stop the transaction.

File: smartcontracts/CryptaurDepository.sol (hunk 3 above)
Comment (soloveyhappy) [reply smartcontracts/CryptaurDepository.sol]: fix in next commit

File: smartcontracts/CryptaurDepository.sol (hunk 2 above)
Comment (soloveyhappy) [reply smartcontracts/CryptaurDepository.sol]: fix in next commit

File: smartcontracts/CryptaurDepository.sol (hunk 1 above)
Comment (soloveyhappy) [reply smartcontracts/CryptaurDepository.sol]: +

File: smartcontracts/CryptaurDepository.sol [hunk 6]
```diff
@@ -107,96 +227,130 @@ contract CryptaurDepository is owned {
...
             }
         }
         return avail;
     }
     
     // _opinionLeader is optional, can be zero
     function pay(address _seller, uint _amount, address _opinionLeader) public {
-        require(balanceOf[msg.sender] >= _amount);
+        require(balances[msg.sender] >= _amount);
+        address sellerMasterWallet = getOrAddMasterWallet(_seller);
```
[the remark this answers is not in the document]
Comment (soloveyhappy) [reply smartcontracts/CryptaurDepository.sol]: Here we translate to masterWallet only users addresses. In that case msg.sender is the contract address.

[the remark this answers is not in the document]
Comment (isvirin) [reply smartcontracts/CryptaurDepository.sol]: No, it is not truth. Pay can be called by user.

**Pull request merged**